app/src/main/java/org/mozilla/fenix/settings/sitepermissions/SitePermissionsExceptionsFragment.kt

```diff
@@ -58,25 +60,32 @@ class SitePermissionsExceptionsFragment :
         recyclerView = rootView.findViewById(R.id.exceptions)
         recyclerView.layoutManager = LinearLayoutManager(requireContext())
 
-        viewLifecycleOwner.lifecycleScope.launch(Dispatchers.IO) {
-            val sitePermissionsPaged = requireContext().components.core.permissionStorage.getSitePermissionsPaged()
-
-            withContext(Main) {
-                val adapter = ExceptionsAdapter(this@SitePermissionsExceptionsFragment)
-                val liveData = LivePagedListBuilder(sitePermissionsPaged, MAX_ITEMS_PER_PAGE).build()
-
-                liveData.observe(
-                    viewLifecycleOwner,
-                    {
-                        if (it.isEmpty()) {
-                            showEmptyListMessage()
-                        } else {
-                            hideEmptyListMessage()
-                            adapter.submitList(it)
-                            recyclerView.adapter = adapter
-                        }
+        val adapter = ExceptionsAdapter(this).apply {
+            addLoadStateListener { loadState ->
+                if (loadState.source.refresh is LoadState.NotLoading &&
+                    loadState.append.endOfPaginationReached &&
+                    itemCount < 1
+                ) {
+                    showEmptyListMessage()
+                } else {
+                    if (itemCount != 0) {
```

Comment:
Can we convert this to an appropriate else if conditional? Please retest afterwards.